### PR TITLE
bpo-37579: Return NotImplemented in Python implementation of __eq__ to match C implementation of datetime for timedelta and time

### PR DESCRIPTION
**Pull request by tirkarthi**
When the other object being compared is not the same type then :exc:`NotImplemented` is returned in C implementation of `datetime.timedelta` and `datetime.time`. Change the Python implementation to match the C implementation

https://bugs.python.org/issue37579

### CODE REVIEW

**Review by jdemeyer — APPROVED** [review]
Code makes sense, test added.

File: Lib/test/datetimetester.py [hunk 1]
```diff
@@ -399,6 +400,11 @@ def test_harmless_mixed_comparison(self):
         self.assertIn(me, [1, 20, [], me])
         self.assertIn([], [me, 1, 20, []])
 
+        # Comparison over a type other that is not object's type should return
+        # NotImplemented and fallback to other.__eq__. In this case ANY.__eq__
+        # always returns True.
+        self.assertEqual(me, ANY)
```
Comment (pganssle) [inline Lib/test/datetimetester.py]: Is this a guaranteed property of `mock.ANY`? It doesn't seem to be [documented here](https://docs.python.org/3/library/unittest.mock.html#any).

In `dateutil` I have a [dedicated test object for this](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py#L236). If "is an object that compares equal to everything" is not a documented property of `mock.ANY`, it might be worth just creating a test object specifically for this purpose.

File: Lib/test/datetimetester.py [hunk 2]
```diff
@@ -399,6 +400,11 @@ def test_harmless_mixed_comparison(self):
         self.assertIn(me, [1, 20, [], me])
         self.assertIn([], [me, 1, 20, []])
 
+        # Comparison over a type other that is not object's type should return
+        # NotImplemented and fallback to other.__eq__. In this case ANY.__eq__
+        # always returns True.
```
Comment (pganssle) [inline Lib/test/datetimetester.py]: I find this comment a bit hard to parse (specifically the "Comparison over a type other"...), maybe this instead?

```python
        # Comparison to objects of unsupported types should return
        # NotImplemented which falls back to the right hand side's __eq__
        # method. In this case, ANY.__eq__ always returns True.
```

File: Lib/test/datetimetester.py (hunk 1 above)
Comment (pganssle) [inline Lib/test/datetimetester.py]: I also think it would be a good idea to check negative comparison (as is done earlier in this test), because the fallback path depends on which methods are implemented. See [this test in `dateutil`](https://github.com/dateutil/dateutil/blob/master/dateutil/test/test_tz.py#L714), for example.

**Review by pganssle — CHANGES_REQUESTED** [review]
Thank you for doing this @tirkarthi. I have a few minor quibbles inline, but overall this is a very solid PR.

File: Lib/test/datetimetester.py (hunk 1 above)
Comment (tirkarthi) [reply Lib/test/datetimetester.py]: `mock.ANY` is to be used to in the context of mock helper comparisons where always returns True for all objects. So it is like a copy of the test object used in dateutil. I guess I will just create a similar one for the test and don't want to use undocumented usage of being it outside mock helpers contradicting myself https://bugs.python.org/issue37555#msg347731 :)

https://github.com/python/cpython/blob/994a3b88dca852696351358e2743313e546b5ecf/Lib/unittest/mock.py#L2263-L2270

**Comment by tirkarthi** [comment]
@pganssle Thank you for the review. I am convinced to use a dedicated test object instead of `mock.ANY`. I copied the test class from dateutil. I reworded the code comment along with a test for `__ne__`. I have made the requested changes; please review again

**Review by pganssle — APPROVED** [review]

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by pganssle** [comment]
Thanks @tirkarthi for the PR and @serhiy-storchaka for the review and report.

**Comment by tirkarthi** [comment]
Thanks Paul. Is this worth backporting? this seems like a bug fix to me.

**Comment by serhiy-storchaka** [comment]
This seems like a bug fix to me too.

**Comment by pganssle** [comment]
I'm fine with backporting it since it's a PEP 399 violation.
